### PR TITLE
separateDebugInfo: Use --strip-unneeded

### PR DESCRIPTION
**Pull request by andersk**
###### Motivation for this change

According to https://stackoverflow.com/q/46197810/115030, `--only-keep-debug` preserves all the information stripped by `--strip-unneeded`. This reduces the size of the webkitgtk output by 22% (123 MB → 96 MB).

Inspired by #159612.

Not sure whether to target `staging` or `master`. I expect this to touch ~~a lot of packages~~ okay, just about every package (via python, which has `separateDebugInfo = true`), so I went with `staging` for now. But on the other hand `master` is currently blocked from feeding `nixos-unstable` due to the GNOME ISO size issue.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by andersk** [comment]
This seems to break the Valgrind tests in `libpsl`, so needs more investigation.

**Comment by andersk** [comment]
That failure can be reproduced with `valgrind true`. In particular, Valgrind looks for `/usr/lib/debug/.build-id/b2/349b9bd14cc4da986c66895d5a6f238e4de2f9.debug`, which doesn’t exist, and doesn’t look at `/nix/store/zfspvsci7i0a953r7i5b44yfdnwbm0cw-glibc-2.33-108-debug/lib/debug/.build-id/b2/349b9bd14cc4da986c66895d5a6f238e4de2f9.debug` even with `--extra-debuginfo-path=/nix/store/zfspvsci7i0a953r7i5b44yfdnwbm0cw-glibc-2.33-108-debug/lib/debug`. Here’s a matching [upstream bug](https://bugs.kde.org/show_bug.cgi?id=386602).

**Comment by andersk** [comment]
I guess we’ll need to patch Valgrind to respect `NIX_DEBUG_INFO_DIRS`, like we did for GDB and elfutils.

File: pkgs/development/libraries/libpsl/default.nix [hunk 1]
```diff
@@ -38,6 +39,7 @@ in stdenv.mkDerivation rec {
     python3
     libxslt
   ] ++ lib.optionals enableValgrindTests [
+    glibc.debug
```
Comment (andersk) [inline pkgs/development/libraries/libpsl/default.nix:42]: (Should this be in valgrind’s `propagatedBuildInputs`, or be left up to its users?)

**Review by andersk — COMMENTED** [review]
Patched Valgrind, and now I can build the entire closure of webkitgtk.

**Review by c0bw3b — APPROVED** [review]
Ok with the change on `libpsl`

Result of `nixpkgs-review pr 160259` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libpsl</li>
  </ul>
</details>

---

As for `--strip-unneeded` it seems safe indeed.
Do we have a rough estimate of size difference (and thus the space that will be saved) compared to `--strip-debug`?

File: pkgs/development/libraries/libpsl/default.nix (hunk 1 above)
Comment (c0bw3b) [reply pkgs/development/libraries/libpsl/default.nix:42]: Good question...!
I'm not sure every use-case of `valgrind` (in and out of nixpkgs) would expect `glibc` debug infos to be available. Maybe. 🤷

File: pkgs/build-support/setup-hooks/separate-debug-info.sh [hunk 2]
```diff
@@ -34,7 +34,7 @@ _separateDebugInfo() {
         # firmware blobs in QEMU.)
         (
             $OBJCOPY --only-keep-debug "$i" "$dst/${id:0:2}/${id:2}.debug"
-            $STRIP --strip-debug "$i"
+            $STRIP --strip-unneeded "$i"
```
Comment (sternenseemann) [inline pkgs/build-support/setup-hooks/separate-debug-info.sh:37]: Will this ever be used on macOS? FWIW it supports neither flags.

strip is an area I've noticed where we tend to use non portable flags, maybe it's time to introduce a more bespoke strip wrapper…

Comment (vcunat) [reply pkgs/build-support/setup-hooks/separate-debug-info.sh:37]: OK, all this sounds more like future improvements.

**Comment by vcunat** [comment]
A more interesting thing would be to switch the default strip flag, i.e. even packages not using `separateDebugInfo`, but right now I didn't feel confident enough to do that.

**Comment by vcunat** [comment]
Uh, this PR is most likely what broke elfutils test(s).  No idea why, but I can reproduce the issue :-/
https://hydra.nixos.org/log/31vfs30d52df2lkfjjd9b048b47f4m60-elfutils-0.186.drv

**Comment by trofi** [comment]
Thread iterators rely on superfluous symbols to be retained in `libpthread.so.0` (`glibc` provides the mechanism itself): https://sourceware.org/gdb/wiki/FAQ#GDB_does_not_see_any_threads_besides_the_thread_in_which_crash_occurred.3B_or_SIGTRAP_kills_my_program_when_I_set_a_breakpoint.

I suspect we over-strip `glibc`.

**Comment by vcunat** [comment]
"unneeded" :-(

So I suppose we'll be reverting this PR, at least for now?  (fixing glibc will be a stdenv rebuild anyway)

**Comment by trofi** [comment]
Also, how should the tests look like in case of separate debug into? Looks like elfutils just tries to get a symbolised backtrace out of crashes binary and expects libc symbols from there. From what I understand `debug` output is not pulled in by default.

What is the correct way to pull it in in libc-agnostic way? AFAIU `glibc.debug` import sounds like a very narrow way to do it (does not work for `musl` alternative), `stdenv.libc.debug` seems to single out libcs without separate debug output.

**Comment by vcunat** [comment]
I expect you rarely need to pull it _into a build_.  For real-life usage it might be most practical to use [`dwarffs`](https://github.com/edolstra/dwarffs).

**Pull request merged**